### PR TITLE
allow serializable objects to bypass full attributes hash construction...

### PR DESCRIPTION
**Pull request by johncronan**
...by supplying an attribute_names method (as activerecord does).

When doing something like:
    relation.as_json except: [:expensive_field]

where expensive_field is, say, a large JSON object in Postgres, we still have to incur the expense of decoding that field, because serializable_hash calls attributes.keys, which builds the hash and then discards its values.

This patch makes serializable_hash check first to see if there is an attribute_names method, such as the one provided by ActiveRecord, and if so it will use that instead.

### CODE REVIEW

**Comment by carlosantoniodasilva** [comment]
I'd rather not couple that with a method provided by Active Record, since it's an optimisation that seems to only make sense in that context.

Maybe we could extract the `attributes.keys` call into a hook method like `attributes_for_serialization`, that Active Record could override to return whichever implementation makes more sense there (perf-wise speaking), like `attribute_names`?

Thoughts?

**Comment by sgrif** [comment]
Isn't `attribute_names` provided by ActiveModel, too?

**Comment by johncronan** [comment]
@carlosantoniodasilva, sounds good to me. Does the above commit do it alright?

@sgrif, no, not that I can see at least.

**Comment by sgrif** [comment]
I think we should go with @carlosantoniodasilva's suggestion, but have the method name be `attribute_names`, with `attributes.keys` as the implementation supplied by Active Model.

**Comment by johncronan** [comment]
Okay, trying that against the CI now.

**Comment by johncronan** [comment]
Okay, CI passes. Any thoughts on merging it in? Either this commit or 9ae1883.

It's at least a bit of a performance win for anyone using :only and :except with as_json, where large tables or expensive fields are involved.

File: activemodel/lib/active_model/serialization.rb [hunk 1]
```diff
@@ -120,6 +120,13 @@ def serializable_hash(options = nil)
       hash
     end
 
+    # Hook method for defining the set of available attribute names. Override
+    # this method in subclasses where, e.g., it's possible to know the names
+    # without having to make an attributes hash.
+    def attribute_names
+      attributes.keys
+    end
```
Comment (bf4) [inline activemodel/lib/active_model/serialization.rb]: Interestingly, [I just implemented something like this in ActiveModelSerializers::Model](https://github.com/rails-api/active_model_serializers/blob/0422a1e7724023aa4c54b7fee683660641d7d523/lib/active_model_serializers/model.rb#L9-L43) to use attribute_names, and ran into a bug where my 'attribute_names' was conflicting with the one in ActiveModel::Serialization (which led me here).  Also, I got the idea from Crafting Rails Applications by Jose Valim, one of the [original authors of this part of the code and AMS](https://github.com/rails-api/active_model_serializers/blob/0422a1e7724023aa4c54b7fee683660641d7d523/CHANGELOG.md#prehistory). So, good provenance.

**Comment by kamipo** [comment]
Closing in favor of #36059.

**Pull request closed without merging**